packages/ra-core/src/i18n/TestTranslationProvider.tsx

```diff
@@ -14,12 +14,15 @@ export const TestTranslationProvider = ({
     </I18nContextProvider>
 );
 
+export interface IMessages
+    extends Record<string, string | ((options?: any) => string) | IMessages> {}
+
 export const testI18nProvider = ({
     translate,
     messages,
 }: {
     translate?: I18nProvider['translate'];
-    messages?: Record<string, string | ((options?: any) => string)>;
+    messages?: IMessages;
```

Comment:
Had to make this change because otherwise TypeScript (rightfully) complains when `messages` contains nested keys.

Example:

```js
messages: {
  resources: {
    posts: {
      title: 'Post'
    }
  }
}
```